File: csharp/compatibility_tests/v3.0.0/src/Google.Protobuf.Test/WellKnownTypes/DurationTest.cs
#region Copyright notice and license
// Protocol Buffers - Google's data interchange format
// Copyright 2015 Google Inc.  All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd
#endregion

using NUnit.Framework;
using System;

namespace Google.Protobuf.WellKnownTypes
{
    public class DurationTest
    {
        [Test]
        public void ToTimeSpan()
        {
            Assert.AreEqual(TimeSpan.FromSeconds(1), new Duration { Seconds = 1 }.ToTimeSpan());
            Assert.AreEqual(TimeSpan.FromSeconds(-1), new Duration { Seconds = -1 }.ToTimeSpan());
            Assert.AreEqual(TimeSpan.FromMilliseconds(1), new Duration { Nanos = 1000000 }.ToTimeSpan());
            Assert.AreEqual(TimeSpan.FromMilliseconds(-1), new Duration { Nanos = -1000000 }.ToTimeSpan());
            Assert.AreEqual(TimeSpan.FromTicks(1), new Duration { Nanos = 100 }.ToTimeSpan());
            Assert.AreEqual(TimeSpan.FromTicks(-1), new Duration { Nanos = -100 }.ToTimeSpan());

            // Rounding is towards 0
            Assert.AreEqual(TimeSpan.FromTicks(2), new Duration { Nanos = 250 }.ToTimeSpan());
            Assert.AreEqual(TimeSpan.FromTicks(-2), new Duration { Nanos = -250 }.ToTimeSpan());
        }

        [Test]
        public void Addition()
        {
            Assert.AreEqual(new Duration { Seconds = 2, Nanos = 100000000 },
                new Duration { Seconds = 1, Nanos = 600000000 } + new Duration { Nanos = 500000000 });
            Assert.AreEqual(new Duration { Seconds = -2, Nanos = -100000000 },
                new Duration { Seconds = -1, Nanos = -600000000 } + new Duration { Nanos = -500000000 });
            Assert.AreEqual(new Duration { Seconds = 1, Nanos = 100000000 },
                new Duration { Seconds = 1, Nanos = 600000000 } + new Duration { Nanos = -500000000 });

            // Non-normalized durations, or non-normalized intermediate results
            Assert.AreEqual(new Duration { Seconds = 1 },
                new Duration { Seconds = 1, Nanos = -500000000 } + new Duration { Nanos = 500000000 });

            Assert.AreEqual(new Duration { Nanos = -900000000 },
                new Duration { Seconds = -1, Nanos = -100000000 } + new Duration { Nanos = 200000000 });
            Assert.AreEqual(new Duration { Nanos = 900000000 },
                new Duration { Seconds = 1, Nanos = 100000000 } + new Duration { Nanos = -200000000 });
        }

        [Test]
        public void Subtraction()
        {
            Assert.AreEqual(new Duration { Seconds = 1, Nanos = 100000000 },
                new Duration { Seconds = 1, Nanos = 600000000 } - new Duration { Nanos = 500000000 });
            Assert.AreEqual(new Duration { Seconds = -1, Nanos = -100000000 },
                new Duration { Seconds = -1, Nanos = -600000000 } - new Duration { Nanos = -500000000 });
            Assert.AreEqual(new Duration { Seconds = 2, Nanos = 100000000 },
                new Duration { Seconds = 1, Nanos = 600000000 } - new Duration { Nanos = -500000000 });

            // Non-normalized durations
            Assert.AreEqual(new Duration(),
                new Duration { Seconds = 1, Nanos = -500000000 } - new Duration { Nanos = 500000000 });
            Assert.AreEqual(new Duration { Seconds = 1 },
                new Duration { Nanos = 2000000000 } - new Duration { Nanos = 1000000000 });
        }

        [Test]
        public void FromTimeSpan()
        {
            Assert.AreEqual(new Duration { Seconds = 1 }, Duration.FromTimeSpan(TimeSpan.FromSeconds(1)));
            Assert.AreEqual(new Duration { Nanos = Duration.NanosecondsPerTick }, Duration.FromTimeSpan(TimeSpan.FromTicks(1)));
        }

        [Test]
        [TestCase(0, Duration.MaxNanoseconds + 1)]
        [TestCase(0, Duration.MinNanoseconds - 1)]
        [TestCase(Duration.MinSeconds - 1, 0)]
        [TestCase(Duration.MaxSeconds + 1, 0)]
        [TestCase(1, -1)]
        [TestCase(-1, 1)]
        public void ToTimeSpan_Invalid(long seconds, int nanoseconds)
        {
            var duration = new Duration { Seconds = seconds, Nanos = nanoseconds };
            Assert.Throws<InvalidOperationException>(() => duration.ToTimeSpan());
        }

        [Test]
        [TestCase(0, Duration.MaxNanoseconds)]
        [TestCase(0, Duration.MinNanoseconds)]
        [TestCase(Duration.MinSeconds, Duration.MinNanoseconds)]
        [TestCase(Duration.MaxSeconds, Duration.MaxNanoseconds)]
        public void ToTimeSpan_Valid(long seconds, int nanoseconds)
        {
            // Only testing that these values don't throw, unlike their similar tests in ToTimeSpan_Invalid
            var duration = new Duration { Seconds = seconds, Nanos = nanoseconds };
            duration.ToTimeSpan();
        }

        [Test]
        public void ToString_NonNormalized()
        {
            // Just a single example should be sufficient...
            var duration = new Duration { Seconds = 1, Nanos = -1 };
            Assert.AreEqual("{ \"@warning\": \"Invalid Duration\", \"seconds\": \"1\", \"nanos\": -1 }", duration.ToString());
        }
    }
}

File: python/google/protobuf/internal/duration_test.py
# -*- coding: utf-8 -*-
# Protocol Buffers - Google's data interchange format
# Copyright 2008 Google Inc.  All rights reserved.
#
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file or at
# https://developers.google.com/open-source/licenses/bsd

"""Tests proto Duration APIs."""

import datetime
import unittest

from google.protobuf import duration
from google.protobuf.internal import testing_refleaks
from google.protobuf.internal import well_known_types_test_pb2

from google.protobuf import duration_pb2


@testing_refleaks.TestCase
class DurationTest(unittest.TestCase):

  def test_duration_integer_conversion(self):
    self.assertEqual(1, duration.to_nanoseconds(duration.from_nanoseconds(1)))
    self.assertEqual(-1, duration.to_seconds(duration.from_seconds(-1)))
    self.assertEqual(
        123, duration.to_milliseconds(duration.from_milliseconds(123))
    )
    self.assertEqual(
        321, duration.to_microseconds(duration.from_microseconds(321))
    )

  def test_duration_json(self):

    def check_duration(message, text):
      self.assertEqual(text, duration.to_json_string(message))
      parsed_duration = duration.from_json_string(text)
      self.assertEqual(message, parsed_duration)

    message = duration_pb2.Duration()
    message.seconds = 0
    message.nanos = 0
    check_duration(message, '0s')
    message.nanos = 10000000
    check_duration(message, '0.010s')
    message.nanos = 10000
    check_duration(message, '0.000010s')
    message.nanos = 10
    check_duration(message, '0.000000010s')

  def test_duration_timedelta(self):
    message = duration.from_nanoseconds(1999999999)
    td = duration.to_timedelta(message)
    self.assertEqual(1, td.seconds)
    self.assertEqual(999999, td.microseconds)

    message = duration.from_microseconds(-1)
    td = duration.to_timedelta(message)
    converted_message = duration.from_timedelta(td)
    self.assertEqual(message, converted_message)

  def test_duration_construction(self):
    expected_td = datetime.timedelta(microseconds=123)
    message = well_known_types_test_pb2.WKTMessage(
        optional_duration=expected_td
    )
    self.assertEqual(expected_td, message.optional_duration.ToTimedelta())

  def test_duration_sub_annotation(self):
    dt = datetime.datetime.now()
    dr = duration_pb2.Duration()
    td = datetime.timedelta(microseconds=123)
    # datetime - Duration
    self.assertEqual(dt - dr, dt - duration.to_timedelta(dr))
    # timedelta - Duration and Duration - Duration
    self.assertEqual(td - dr, duration.from_timedelta(td) - dr)
    # Duration - timedelta
    self.assertEqual(dr - td, dr - duration.from_timedelta(td))

  def test_duration_add_annotation(self):
    dt = datetime.datetime.now()
    dr = duration_pb2.Duration()
    dr2 = duration_pb2.Duration(seconds=100)
    # datetime + Duration and Duration + datetime
    self.assertEqual(dt + dr, dr + dt)
    message = well_known_types_test_pb2.WKTMessage(optional_timestamp=dt)
    # Duration + Timestamp
    self.assertEqual(dr + message.optional_timestamp, dr + dt)
    td = datetime.timedelta(microseconds=123)
    # Duration + timedelta and timedelta + Duration
    self.assertEqual(dr + td, td + dr)
    # Duration + Duration
    self.assertEqual(dr + dr2, dr2 + dr)


if __name__ == '__main__':
  unittest.main()
